src/Spice86.Core/Emulator/Devices/Input/Keyboard/PS2Keyboard.cs

```diff
@@ -0,0 +1,569 @@
+namespace Spice86.Core.Emulator.Devices.Input.Keyboard;
+
+using Serilog.Events;
+
+using Spice86.Core.Emulator.Devices.ExternalInput;
+using Spice86.Shared.Emulator.Keyboard;
+using Spice86.Shared.Interfaces;
+
+using System.Collections.Generic;
+using System.Diagnostics;
+
+/// <summary>
+/// Emulates a PS/2 keyboard device, handling key events, scancode translation, and communication with a keyboard
+/// controller.
+/// </summary>
+[DebuggerDisplay("PS2Keyboard Set={_codeSet} Scanning={_isScanning} Buf={_bufferCurrentIndex}/{BufferSize} Overflowed={_bufferOverflowed} Repeat={_repeat.Key} WaitMs={_repeat.WaitMs}")]
+public partial class PS2Keyboard {
+    private readonly Intel8042Controller _controller;
+    private readonly ILoggerService _loggerService;
+    private readonly KeyboardScancodeConverter _scancodeConverter = new();
+    private readonly DualPic _dualPic;
+    private readonly IGuiKeyboardEvents? _gui;
+    private readonly EmulatedTimeEventHandler _typematicTickHandler;
+    private readonly EmulatedTimeEventHandler _ledsAllOnExpireHandler;
+
+    /// <summary>
+    /// Internal keyboard scancode buffer
+    /// </summary>
+    private const int BufferSizeInScancodes = 8;
+    private readonly List<byte>[] _buffer = new List<byte>[BufferSizeInScancodes];
```

Comment:
Could this be a csharp queue? Maybe a better name would be _keyboardFrameBuffer?

Reply:
Yes that would be better.

Although looking at the usage, isn't it a circular buffer ?